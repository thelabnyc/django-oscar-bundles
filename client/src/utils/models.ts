import * as t from 'io-ts';
import {failure} from 'io-ts/lib/PathReporter';
import {Either, isLeft} from 'fp-ts/lib/Either';


export const check = <T>(result: Either<t.Errors, T>): T => {
    if (isLeft(result)) {
        throw new Error(failure(result.left).join('\n'));
    }
    return result.right;
};

const nullable = <RT extends t.Mixed>(type: RT) => {
    return t.union([t.null, type]);
};

export const optional = <RT extends t.Mixed>(type: RT) => {
    return t.union([t.undefined, type]);
};

class FileType extends t.Type<File> {
    readonly _tag: 'NumberType' = 'NumberType';
    constructor() {
        super(
            'number',
            (f): f is File => {
                return f instanceof File;
            },
            (u, c) => {
                return this.is(u) ? t.success(u) : t.failure(u, c);
            },
            t.identity
        );
    }
}
const file = new FileType();



export const DRFSelectOption = t.interface({
    value: t.string,
    display_name: t.string,
});
export const DRFSelectOptions = t.readonlyArray(DRFSelectOption);

export const BaseField = <T extends string>(fieldType: t.LiteralC<T>) => {
    return t.interface({
        type: fieldType,
        required: t.boolean,
        read_only: t.boolean,
        label: t.string,
        help_text: optional(t.string),
    });
};

export const IntegerField = BaseField(t.literal("integer"));

export const BooleanField = BaseField(t.literal("boolean"));

export const DatetimeField = BaseField(t.literal("datetime"));

export const StringField = t.intersection([
    BaseField(t.literal("string")),
    t.interface({
        max_length: optional(t.number),
    }),
]);

export const ChoiceField = t.intersection([
    BaseField(t.literal("choice")),
    t.interface({
        choices: DRFSelectOptions,
    }),
]);

export const Field = t.union([
    IntegerField,
    BooleanField,
    DatetimeField,
    StringField,
    ChoiceField,
    t.unknown,
]);

export const HTTPMethod = t.union([
    t.literal("GET"),
    t.literal("HEAD"),
    t.literal("OPTIONS"),
    t.literal("POST"),
    t.literal("PUT"),
    t.literal("PATCH"),
    t.literal("DELETE"),
]);

export const DRFOptionsResponse = t.interface({
    name: t.string,
    description: t.string,
    renders: t.readonlyArray(t.string),
    parses: t.readonlyArray(t.string),
    actions: t.interface({
        GET: optional(t.record(t.string, Field)),
        HEAD: optional(t.record(t.string, Field)),
        OPTIONS: optional(t.record(t.string, Field)),
        POST: optional(t.record(t.string, Field)),
        PUT: optional(t.record(t.string, Field)),
        PATCH: optional(t.record(t.string, Field)),
        DELETE: optional(t.record(t.string, Field)),
    }),
});




export const Product = t.interface({
    id: t.number,
    dashboard_url: t.string,
    product_class: nullable(t.interface({
        id: t.number,
        name: t.string,
    })),
    title: t.string,
    slug: t.string,
    is_parent: t.boolean,
    is_child: t.boolean,
    parent: nullable(t.number),
    children: t.array(t.number),
});
export const Products = t.array(Product);



export const Range = t.interface({
    id: t.number,
    dashboard_url: t.string,
    name: t.string,
    slug: t.string,
    description: t.string,
});
export const Ranges = t.array(Range);



export const ConcreteBundle = t.interface({
    id: nullable(t.number),
    triggering_product: t.number,
    suggested_products: t.array(t.number),
});
export const ConcreteBundles = t.array(ConcreteBundle);



export const UserConfigurableBundle = t.interface({
    id: nullable(t.number),
    triggering_product: t.number,
    suggested_range: t.number,
    quantity: t.number,
});
export const UserConfigurableBundles = t.array(UserConfigurableBundle);



export const BundleGroup = t.interface({
    id: nullable(t.number),
    bundle_type: t.string,
    name: t.string,
    description: t.string,
    headline: t.string,
    image: optional(nullable(t.string)),
    newImage: optional(nullable(file)),
    clearImage: optional(t.boolean),
    triggering_parents: t.array(t.number),
    suggested_parents: t.array(t.number),
    concrete_bundles: t.array(ConcreteBundle),
    user_configurable_bundles: t.array(UserConfigurableBundle),
});
export const BundleGroups = t.array(BundleGroup);
