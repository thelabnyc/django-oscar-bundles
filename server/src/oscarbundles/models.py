from django.db import models
from django.conf import settings
from django.utils.translation import ugettext_lazy as _
from oscar.core.loading import get_model

Product = get_model('catalogue', 'Product')


class BundleGroup(models.Model):
    GROUP_TYPE_OPTIONS = getattr(settings, 'BUNDLE_GROUP_TYPES', (
        ('default', 'Default'),
    ))

    bundle_type = models.CharField(_('Bundle Type'),
        max_length=200,
        default=GROUP_TYPE_OPTIONS[0][0],
        choices=GROUP_TYPE_OPTIONS)

    name = models.CharField(_('Name'),
        max_length=200,
        blank=True,
        default='')
    headline = models.CharField(
        max_length=200,
        blank=True,
        default='Forget Something?',
        help_text='CTA headline in cart display')
    description = models.TextField(_('Description'),
        blank=True,
        default='')
    image = models.ImageField(
        upload_to=settings.OSCAR_IMAGE_FOLDER,
        max_length=255,
        blank=True,
        null=True)

    triggering_parents = models.ManyToManyField('catalogue.Product',
        related_name='triggering_bundle_groups',
        verbose_name=_('Triggering Products'))
    suggested_parents = models.ManyToManyField('catalogue.Product',
        related_name='suggesting_bundle_groups',
        verbose_name=_('Suggested Products'))


class Bundle(models.Model):
    bundle_group = models.ForeignKey(BundleGroup,
        on_delete=models.CASCADE,
        related_name='bundles')

    triggering_product = models.ForeignKey('catalogue.Product',
        on_delete=models.CASCADE,
        related_name='triggering_bundles',
        verbose_name=_('Triggering Product'),
        help_text=_('Which product should trigger this bundle?'))
    suggested_products = models.ManyToManyField('catalogue.Product',
        related_name='suggesting_bundles',
        verbose_name=_('Suggested Products'),
        help_text=_('Which product(s) should this bundle suggest when triggered?'))

    class Meta:
        unique_together = (
            ('bundle_group', 'triggering_product'),
        )
