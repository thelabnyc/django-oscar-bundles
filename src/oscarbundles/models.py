from django.db import models
from django.conf import settings
from django.utils.translation import ugettext_lazy as _
from oscar.core.loading import get_model

Product = get_model('catalogue', 'Product')


class BundleMetaMixin(models.Model):
    name = models.CharField(_('Name'),
        max_length=200,
        blank=True,
        default='',
        help_text=_('Optional name for this bundle'))
    description = models.TextField(_('Description'),
        blank=True,
        default='',
        help_text=_('Optional description for this bundle'))
    image = models.ImageField(
        upload_to=settings.OSCAR_IMAGE_FOLDER, max_length=255,
        blank=True, null=True)

    class Meta:
        abstract = True

    def __str__(self):
        return self.name


class Bundle(BundleMetaMixin, models.Model):
    bundle_group = models.ForeignKey('BundleGroup',
        blank=True, null=True, help_text=_('Optional grouping for bundle'), on_delete=models.CASCADE)
    triggering_products = models.ManyToManyField('catalogue.Product',
        related_name='triggering_bundles',
        verbose_name=_('Triggering Products'),
        help_text=_('Which product(s) should trigger this bundle?'))
    suggested_products = models.ManyToManyField('catalogue.Product',
        related_name='suggesting_bundles',
        verbose_name=_('Suggested Products'),
        help_text=_('Which product(s) should this bundle suggest when triggered?'))


class BundleGroup(BundleMetaMixin, models.Model):
    pass
