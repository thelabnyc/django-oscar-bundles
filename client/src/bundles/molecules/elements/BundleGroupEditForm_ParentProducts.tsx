import React from "react";
import classNames from "classnames";
import Select from "react-select";
import { IProduct, SelectOption } from "../../../utils/models.interfaces";

interface IProps {
    products: IProduct[];
    isSaving: boolean;
    errors: {
        triggering_parents?: string[];
        suggested_parents?: string[];
    };
    triggeringParents: number[];
    suggestedParents: number[];
    onSelectParent: (
        name: "triggeringParents" | "suggestedParents",
        opts: SelectOption | ReadonlyArray<SelectOption> | null | undefined,
    ) => void;
}

interface IState {}

export class ParentProductsEditForm extends React.PureComponent<
    IProps,
    IState
> {
    private getParentProductSelectOptions() {
        return this.props.products
            .filter((p) => {
                return !p.is_child;
            })
            .sort((a, b) => {
                return a.title.localeCompare(b.title);
            })
            .map((p) => {
                const opt: SelectOption = {
                    value: p.id,
                    label: p.title,
                };
                return opt;
            });
    }

    private buildFormGroupClasses(
        field: keyof IProps["errors"],
        extras: string[],
    ) {
        const fieldErrors = this.props.errors[field];
        const classes: { [name: string]: boolean } = {
            "form-group": true,
            "has-error": fieldErrors ? fieldErrors.length > 0 : false,
        };
        extras.forEach((className) => {
            classes[className] = true;
        });
        return classNames(classes);
    }

    private buildErrors(field: keyof IProps["errors"]) {
        return (this.props.errors[field] || []).map((errorMsg, i) => {
            return (
                <span key={i} className="help-block">
                    <i className="fas fa-exclamation"></i> {errorMsg}
                </span>
            );
        });
    }

    render() {
        const parentProductSelectOptions = this.getParentProductSelectOptions();
        return (
            <div className="col-sm-12 bundle-group-edit__section">
                <h1>{gettext("Relevant Products")}</h1>
                <p>
                    <em>
                        {gettext(
                            "Select the parent and standalone products revelant to this bundle group.",
                        )}
                    </em>
                </p>
                <div className="row">
                    <div
                        className={this.buildFormGroupClasses(
                            "triggering_parents",
                            ["col-sm-6"],
                        )}
                    >
                        <label htmlFor="id_image" className="control-label">
                            {gettext("Triggers")}
                        </label>
                        <div>
                            <Select
                                name="triggeringParents"
                                isMulti={true}
                                value={parentProductSelectOptions.filter((o) =>
                                    this.props.triggeringParents.includes(
                                        o.value,
                                    ),
                                )}
                                onChange={this.props.onSelectParent.bind(
                                    this,
                                    "triggeringParents",
                                )}
                                options={parentProductSelectOptions}
                                isDisabled={this.props.isSaving}
                            />
                            {this.buildErrors("triggering_parents")}
                        </div>
                    </div>
                    <div
                        className={this.buildFormGroupClasses(
                            "suggested_parents",
                            ["col-sm-6"],
                        )}
                    >
                        <label htmlFor="id_image" className="control-label">
                            {gettext("Suggestions")}
                        </label>
                        <div>
                            <Select
                                name="suggestedProducts"
                                isMulti={true}
                                value={parentProductSelectOptions.filter((o) =>
                                    this.props.suggestedParents.includes(
                                        o.value,
                                    ),
                                )}
                                onChange={this.props.onSelectParent.bind(
                                    this,
                                    "suggestedParents",
                                )}
                                options={parentProductSelectOptions}
                                isDisabled={this.props.isSaving}
                            />
                            {this.buildErrors("suggested_parents")}
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}
