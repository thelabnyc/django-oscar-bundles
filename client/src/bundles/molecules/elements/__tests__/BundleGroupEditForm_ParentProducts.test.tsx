import React from "react";
import { render } from "@testing-library/react";
import { IProduct } from "../../../../utils/models.interfaces";
import { ParentProductsEditForm } from "../BundleGroupEditForm_ParentProducts";

const products: IProduct[] = [
    {
        id: 1,
        dashboard_url: "http://localhost/store/dashboard/catalogue/products/1/",
        product_class: {
            id: 1,
            name: "books",
        },
        title: "book-1",
        slug: "book-1",
        is_parent: false,
        is_child: false,
        parent: null,
        children: [],
    },
    {
        id: 2,
        dashboard_url: "http://localhost/store/dashboard/catalogue/products/2/",
        product_class: {
            id: 1,
            name: "books",
        },
        title: "book-2",
        slug: "book-2",
        is_parent: false,
        is_child: false,
        parent: null,
        children: [],
    },
];

describe("#bundles/molecules/elements/ParentProductsEditForm", () => {
    it("renders", () => {
        const wrapper = render(
            <ParentProductsEditForm
                products={products}
                isSaving={false}
                errors={{}}
                triggeringParents={[1]}
                suggestedParents={[2]}
                onSelectParent={() => null}
            />,
        );
        expect(wrapper.asFragment()).toMatchSnapshot();
    });

    it("renders error messages", () => {
        const wrapper = render(
            <ParentProductsEditForm
                products={products}
                isSaving={false}
                errors={{
                    triggering_parents: ["Test error message"],
                    suggested_parents: [],
                }}
                triggeringParents={[1]}
                suggestedParents={[2]}
                onSelectParent={() => null}
            />,
        );
        expect(wrapper.asFragment()).toMatchSnapshot();
    });
});
