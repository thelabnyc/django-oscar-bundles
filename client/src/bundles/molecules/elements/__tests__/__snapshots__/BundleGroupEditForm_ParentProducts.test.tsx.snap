// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#bundles/molecules/elements/ParentProductsEditForm renders 1`] = `
<div
  className="col-sm-12 bundle-group-edit__section"
>
  <h1>
    Relevant Products
  </h1>
  <p>
    <em>
      Select the parent and standalone products revelant to this bundle group.
    </em>
  </p>
  <div
    className="row"
  >
    <div
      className="form-group col-sm-6"
    >
      <label
        className="control-label"
        htmlFor="id_image"
      >
        Triggers
      </label>
      <div>
        <StateManager
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          isDisabled={false}
          isMulti={true}
          name="triggeringParents"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "book-1",
                "value": 1,
              },
              Object {
                "label": "book-2",
                "value": 2,
              },
            ]
          }
          value={
            Array [
              Object {
                "label": "book-1",
                "value": 1,
              },
            ]
          }
        />
      </div>
    </div>
    <div
      className="form-group col-sm-6"
    >
      <label
        className="control-label"
        htmlFor="id_image"
      >
        Suggestions
      </label>
      <div>
        <StateManager
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          isDisabled={false}
          isMulti={true}
          name="suggestedProducts"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "book-1",
                "value": 1,
              },
              Object {
                "label": "book-2",
                "value": 2,
              },
            ]
          }
          value={
            Array [
              Object {
                "label": "book-2",
                "value": 2,
              },
            ]
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`#bundles/molecules/elements/ParentProductsEditForm renders error messages 1`] = `
<div
  className="col-sm-12 bundle-group-edit__section"
>
  <h1>
    Relevant Products
  </h1>
  <p>
    <em>
      Select the parent and standalone products revelant to this bundle group.
    </em>
  </p>
  <div
    className="row"
  >
    <div
      className="form-group has-error col-sm-6"
    >
      <label
        className="control-label"
        htmlFor="id_image"
      >
        Triggers
      </label>
      <div>
        <StateManager
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          isDisabled={false}
          isMulti={true}
          name="triggeringParents"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "book-1",
                "value": 1,
              },
              Object {
                "label": "book-2",
                "value": 2,
              },
            ]
          }
          value={
            Array [
              Object {
                "label": "book-1",
                "value": 1,
              },
            ]
          }
        />
        <span
          className="help-block"
          key="0"
        >
          <i
            className="icon-exclamation-sign"
          />
           
          Test error message
        </span>
      </div>
    </div>
    <div
      className="form-group col-sm-6"
    >
      <label
        className="control-label"
        htmlFor="id_image"
      >
        Suggestions
      </label>
      <div>
        <StateManager
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          isDisabled={false}
          isMulti={true}
          name="suggestedProducts"
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "book-1",
                "value": 1,
              },
              Object {
                "label": "book-2",
                "value": 2,
              },
            ]
          }
          value={
            Array [
              Object {
                "label": "book-2",
                "value": 2,
              },
            ]
          }
        />
      </div>
    </div>
  </div>
</div>
`;
